opm/simulators/linalg/bda/IncompleteSAI.cpp

```diff
@@ -0,0 +1,324 @@
+#include <iostream>
+
+#include <config.h>
+
+#include <opm/common/OpmLog/OpmLog.hpp>
+#include <opm/common/ErrorMacros.hpp>
+#include <dune/common/timer.hh>
+#include <opm/simulators/linalg/bda/BlockedMatrix.hpp>
+
+#include <opm/simulators/linalg/bda/IncompleteSAI.hpp>
+
+namespace Opm
+{
+namespace Accelerator
+{
+    using Opm::OpmLog;
+    using Dune::Timer;
+
+    template <unsigned int block_size>
+    void IncompleteSAI<block_size>::setParams(int Nb_, int nnzbs_, int verbosity_){
+        this->Nb = Nb_;
+        this->bs = block_size;
+        this->N = bs * Nb_;
+        this->nnzbs = nnzbs_;
+        this->verbosity = verbosity_;
+    }
+
+    template <unsigned int block_size>
+    void IncompleteSAI<block_size>::buildMapping(int *CSRColIndices, int *CSRRowPointers){
+        Dune::Timer t_mapping;
+        std::vector<int> aux(CSRRowPointers, CSRRowPointers + Nb + 1);
+        mapping.resize(nnzbs);
+
+        for(int row = 0; row < Nb; row++){
+            for(int jj = CSRRowPointers[row]; jj < CSRRowPointers[row+1]; jj++){
+                int col = CSRColIndices[jj];
+                int dest = aux[col];
+                mapping[dest] = jj;
+                aux[col]++;
+            }
+        }
```

Comment:
CSR->CSC for symmetric sparsity pattern? Might have been clearer if this would be in its own function with a speaking name.

I know that the assumption of a symmetric sparsity should hold, but would sleep better if we had the option to check this e.g. when a DEBUG define is set. (at least in AMG we had occasions were it did not in the past. Might be resolved, though).